### PR TITLE
[WIP][staking] non-beacon chain relay staking txn

### LINKED ISSUE #3017: [enhancement] staking tx shard agnostic
**Describe the bug**

To make hmy shard-agnostic, if the staking tx is received by a node on a non-beacon chain shard (currently shard #1, #2 or #3), it should broadcast this transaction to shard 0. This prevents awkward situations, such as the validator not being able to use his own node to send staking transactions.

### PR DESCRIPTION
**Pull request by JackyWYX**
## Issue

Closes #3017. When handling staking transaction, non-beacon chain nodes will relay the message of staking transaction to beacon chain. 

Currently doing local test.

### CODE REVIEW

File: node/node.go [hunk 1]
```diff
@@ -256,19 +255,28 @@ func (node *Node) addPendingStakingTransactions(newStakingTxs staking.StakingTra
...
-		utils.Logger().Info().Str("Hash", newStakingTx.Hash().Hex()).Msg("Broadcasting Staking Tx")
-		node.tryBroadcastStaking(newStakingTx)
 	}
+	utils.Logger().Info().Str("Hash", newStakingTx.Hash().Hex()).Msg("Broadcasting Staking Tx")
+	node.tryBroadcastStaking(newStakingTx)
 	return nil
 }
 
+func (node *Node) isPreStakingEpoch() bool {
+	return node.Blockchain().Config().IsPreStaking(node.Blockchain().CurrentHeader().Epoch())
```
Comment (Daniel-VDM) [inline node/node.go:277]: Would it be better to use Beaconchain?

File: node/node.go [hunk 2]
```diff
@@ -234,9 +234,8 @@ func (node *Node) addPendingTransactions(newTxs types.Transactions) []error {
 
 // Add new staking transactions to the pending staking transaction list.
 func (node *Node) addPendingStakingTransactions(newStakingTxs staking.StakingTransactions) []error {
-	if node.NodeConfig.ShardID == shard.BeaconChainShardID &&
-		node.Blockchain().Config().IsPreStaking(node.Blockchain().CurrentHeader().Epoch()) {
-		poolTxs := types.PoolTransactions{}
+	if node.NodeConfig.ShardID == shard.BeaconChainShardID && node.isPreStakingEpoch() {
+		poolTxs := make(types.PoolTransactions, 0, len(newStakingTxs))
```
Comment (fxfactorial) [inline node/node.go:238]: Do not preallocate . This is wrong premature optimization

File: node/node.go [hunk 3]
```diff
@@ -256,19 +255,28 @@ func (node *Node) addPendingStakingTransactions(newStakingTxs staking.StakingTra
 func (node *Node) AddPendingStakingTransaction(
 	newStakingTx *staking.StakingTransaction,
 ) error {
+	if !node.isPreStakingEpoch() {
+		utils.Logger().Info().Str("Hash", newStakingTx.Hash().Hex()).
```
Comment (fxfactorial) [inline node/node.go:259]: There already is a .hex log thing I think

File: node/node.go (hunk 1 above)
Comment (fxfactorial) [reply node/node.go:277]: Need to use beaconchain

Comment (JackyWYX) [reply node/node.go:277]: Sorry don't quite get it. New logic for `AddPendingStakingTransaction`:

```
if is not pre staking:
    do nothing
elif node is on beacon chain:
    Add to local tx pool
    broadcast staking tx
elif not on beacon chain:
    broadcast staking tx to beacon chain
```

Is there any logic wrong in this pseudocode?

File: node/node.go (hunk 2 above)
Comment (JackyWYX) [reply node/node.go]: Done. But could you tell me a bit about why it's wrong?

File: node/node.go (hunk 3 above)
Comment (JackyWYX) [reply node/node.go:259]: This is newly added log. In previous logic, a node not on beacon chain / not in staking epoch received staking transaction will silently return a nil error. A logging only happens when the node is broadcasting the txn. So is it true that this info doesn't have enough value to be logged?

File: node/node.go (hunk 1 above)
Comment (Daniel-VDM) [reply node/node.go:277]: epochs are only defined by the beacon chain for staking era. `node.Blockchain()` can be a shard chain and thus be wrong.

File: node/node.go (hunk 2 above)
Comment (fxfactorial) [reply node/node.go]: You are trying to be smarter than compiler when it’s not necessary and preallocate lots of space bc this is txn pool which could potentially be big. Append is heavily optimized and pretty damn good on own.

Comment (JackyWYX) [reply node/node.go]: I am sorry I am little bit confused. According to my knowledge, append is done through double extending. In each double extending, a new sized array is allocated. So suppose txs size is `n` and it is big. If we allocate at the beginning, we allocate `n` pointers total. But if we do not use append, we allocate `2n` pointers in total, which is a waste in performace. Am I understanding this correctly?

File: node/node.go (hunk 3 above)
Comment (fxfactorial) [reply node/node.go:259]: No you aren’t. You don’t need to make these kinds of assumptions in a garbage collected language and it’s not necessary to think this deeply over this. Focus on business logic please

File: node/node.go (hunk 2 above)
Comment (fxfactorial) [reply node/node.go]: Member
No you aren’t. You don’t need to make these kinds of assumptions in a garbage collected language and it’s not necessary to think this deeply over this. Focus on business logic please

Comment (JackyWYX) [reply node/node.go]: Is there any related reading materials for recommendation for this topic?

File: node/node.go (hunk 3 above)
Comment (JackyWYX) [reply node/node.go]: Removed

**Comment by JackyWYX** [comment]
The feature is already implemented in go-sdk. Closing.

**Pull request closed without merging**